### ISSUE TITLE
Upgrade error 9.8.1 to 9.9.0: Could not load file or assembly 'System.Buffers' or one of its dependencies.

### ISSUE THREAD

**Issue by idasbv** [dnnsoftware/Dnn.Platform#4525]
upgrade DNN 9.8.1 --> DNN 9.9.0

Could not load file or assembly 'System.Buffers, Version=4.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51' or one of its dependencies. The located assembly's manifest definition does not match the assembly reference. (Exception from HRESULT: 0x80131040)]
   System.Reflection.RuntimeAssembly._nLoad(AssemblyName fileName, String codeBase, Evidence assemblySecurity, RuntimeAssembly locationHint, StackCrawlMark& stackMark, IntPtr pPrivHostBinder, Boolean throwOnFileNotFound, Boolean forIntrospection, Boolean suppressSecurityChecks) +0
   System.Reflection.RuntimeAssembly.InternalLoadAssemblyName(AssemblyName assemblyRef, Evidence assemblySecurity, RuntimeAssembly reqAssembly, StackCrawlMark& stackMark, IntPtr pPrivHostBinder, Boolean throwOnFileNotFound, Boolean forIntrospection, Boolean suppressSecurityChecks) +232
   System.Reflection.RuntimeAssembly.InternalLoad(String assemblyString, Evidence assemblySecurity, StackCrawlMark& stackMark, IntPtr pPrivHostBinder, Boolean forIntrospection) +113
   System.Reflection.RuntimeAssembly.InternalLoad(String assemblyString, Evidence assemblySecurity, StackCrawlMark& stackMark, Boolean forIntrospection) +23
   System.Reflection.Assembly.Load(String assemblyString) +35
   System.Web.Configuration.CompilationSection.LoadAssemblyHelper(String assemblyName, Boolean starDirective) +49

## Affected version

* [ ] 10.00.00 alpha build
* [ ] 09.09.00 release candidate
* [ ] 09.08.01 latest supported release

## Affected browser

* [ ] Chrome
* [ ] Firefox
* [ ] Safari
* [ ] Internet Explorer 11
* [ ] Microsoft Edge (Classic)
* [ ] Microsoft Edge Chromium

**Comment by bdukes** [comment]
This error means that there was already a `System.Buffers.dll` in the `bin/` directory with a binding redirect in the `web.config`, and the upgrade overwrote that file with a different version, so that its binding redirect in the `web.config` is for a different version than what's in the `bin/` directory.

The simplest resolution is to replace `System.Buffers.dll` with the version that was there before.

Alternatively, you can update the binding redirect for `System.Buffers.dll` in the `web.config` to match the version from the install package.  That version it 4.0.3.0, so the binding redirect should look like this:

```
      <dependentAssembly xmlns="urn:schemas-microsoft-com:asm.v1">
        <assemblyIdentity name="System.Buffers" publicKeyToken="cc7b13ffcd2ddd51" />
        <bindingRedirect oldVersion="0.0.0.0-32767.32767.32767.32767" newVersion="4.0.3.0" />
      </dependentAssembly>
```

**Comment by idasbv** [comment]
thanks Brian! did the trick... now I have to find out why there already was a System.Buffers.dll in the bin... maybe a third party module?

**Comment by skamphuis** [comment]
> 
> 
> thanks Brian! did the trick... now I have to find out why there already was a System.Buffers.dll in the bin... maybe a third party module?

Most likely, yes. I recently ran into a case where Vanjaro had installed it, but it might be another one in your case, obviously.
Also, please note that DNN 9.9 includes a newer version of that assembly, so you will need to change that bindingRedirect again.

**Comment by Tychodewaard** [comment]
Recently Mitch mentioned that this error should be in the past from 9.9.1 but I just tested and it is still the case.
Removing the lines made the upgrade continue as expected.
This was an install that included Vanjaro btw.

**Comment by bdukes** [comment]
@Tychodewaard can we get some details so we can walk through your scenario?

Are you using "Vanjaro for DNN 1.0.4 RC"? What version of DNN are you coming from?

Are you saying that the install appears to work correctly, and there's only the persona bar error?  Or are you initially seeing the Yellow Screen of Death?

Thanks!

**Comment by Tychodewaard** [comment]
@bdukes Sure.
This DNN 9.8.1 install still had the Vanjaro beta2 (0.5.4)
- First upgrade try: yellow screen of death, suggesting there are systembuffers issues
- tried adjusting the reference to 4.6 but that did not do anything.
- deleted the reference to systembuffers from the web.config: the upgrade went through
- In the persona bar, the part 'sites' did not work anymore. A quick flash 'sites' and it disappeared again.
- adding back your lines of reference to systembuffers made the 'sites' part reappear in the persona bar

**Comment by bdukes** [comment]
🤦🏻‍♂️ looks like I forgot to exclude `System.Buffers` from the upgrade package.  PR on its way

**Comment by valadas** [comment]
Thanks for spotting that @Tychodewaard and thanks for the quick fix @bdukes It will be included in either the next 9.9.1 RC or in 9.9.1 release.
